log.go

```diff
@@ -0,0 +1,212 @@
+	}
+
+	var template string
+	var message string
+	var parameters []interface{}
+	attrs := map[string]Attribute{}
+
+	for _, arg := range args {
+		switch a := arg.(type) {
+		case string:
+			if template == "" {
+				template = a
+			} else {
+				parameters = append(parameters, a)
+			}
+		case attribute.Builder:
+			attrs[a.Key] = Attribute{
+				Value: a.Value.AsInterface(),
+				Type:  mapTypesToStr[a.Value.Type()],
+			}
+		default:
+			parameters = append(parameters, a)
+		}
+	}
+
+	if template != "" && len(parameters) > 0 {
+		message = fmt.Sprintf(template, parameters...)
+	} else {
+		message = template
+	}
+	if template != "" && len(parameters) > 0 {
+		attrs["sentry.message.template"] = Attribute{
+			Value: template, Type: "string",
+		}
+		for i, p := range parameters {
+			attrs[fmt.Sprintf("sentry.message.parameters.%d", i)] = Attribute{
+				Value: fmt.Sprint(p), Type: "string",
+			}
+		}
+	}
+
+	// handle metadata
+	if release := l.client.options.Release; release != "" {
+		attrs["sentry.release"] = Attribute{Value: release, Type: "string"}
+	}
+	if environment := l.client.options.Environment; environment != "" {
+		attrs["sentry.environment"] = Attribute{Value: environment, Type: "string"}
+	}
+	if serverAddr := l.client.options.ServerName; serverAddr != "" {
+		attrs["sentry.server.address"] = Attribute{Value: serverAddr, Type: "string"}
+	}
+	if spanID.String() != "0000000000000000" {
+		attrs["sentry.trace.parent_span_id"] = Attribute{Value: spanID.String(), Type: "string"}
+	}
+	if sdkIdentifier := l.client.sdkIdentifier; sdkIdentifier != "" {
+		attrs["sentry.sdk.name"] = Attribute{Value: sdkIdentifier, Type: "string"}
+	}
+	if sdkVersion := l.client.sdkVersion; sdkVersion != "" {
+		attrs["sentry.sdk.version"] = Attribute{Value: sdkVersion, Type: "string"}
+	}
+	attrs["sentry.origin"] = Attribute{Value: "auto.logger.log", Type: "string"}
+
+	log := &Log{
+		Timestamp:  time.Now(),
+		TraceID:    traceID,
+		Level:      level,
+		Severity:   severity,
+		Body:       message,
+		Attributes: attrs,
+	}
+	if l.options.BeforeSendLog != nil {
+		log = l.options.BeforeSendLog(log)
+	}
+
+	if log != nil {
+		l.client.batchLogger.logCh <- *log
+	}
+}
+
+func (l *sentryLogger) Trace(ctx context.Context, v ...interface{}) {
+	l.log(ctx, LogLevelTrace, LogSeverityTrace, v...)
+}
+func (l *sentryLogger) Debug(ctx context.Context, v ...interface{}) {
+	l.log(ctx, LogLevelDebug, LogSeverityDebug, v...)
+}
+func (l *sentryLogger) Info(ctx context.Context, v ...interface{}) {
+	l.log(ctx, LogLevelInfo, LogSeverityInfo, v...)
+}
+func (l *sentryLogger) Warn(ctx context.Context, v ...interface{}) {
+	l.log(ctx, LogLevelWarning, LogSeverityWarning, v...)
+}
+func (l *sentryLogger) Error(ctx context.Context, v ...interface{}) {
+	l.log(ctx, LogLevelError, LogSeverityError, v...)
+}
+func (l *sentryLogger) Fatal(ctx context.Context, v ...interface{}) {
+	l.log(ctx, LogLevelFatal, LogSeverityFatal, v...)
+	os.Exit(1)
```

Comment:
why do we exit here? I don't think users would expect their 3rd party logging apis would exit for them.

Reply:
It's actually common in Go that `Fatal()` log method actually executes `os.Exit(1)`. It's on their stdlib anyway.